modules/programs/topgrade.nix

```diff
@@ -0,0 +1,67 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.topgrade;
+
+  tomlFormat = pkgs.formats.toml { };
+
+in {
+
+  meta.maintainers = [ hm.maintainers.msfjarvis ];
+
+  options.programs.topgrade = {
+    enable = mkEnableOption "topgrade";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.topgrade;
+      defaultText = literalExample "pkgs.topgrade";
+      description = "The package to use for the topgrade binary.";
+    };
+
+    settings = mkOption {
+      type = with types;
+        let
+          prim = oneOf [ bool int str ];
+          primOrPrimAttrs = either prim (attrsOf prim);
+          entry = either prim (listOf primOrPrimAttrs);
+          entryOrAttrsOf = t: either entry (attrsOf t);
+          entries = entryOrAttrsOf (entryOrAttrsOf entry);
+        in attrsOf entries // { description = "topgrade configuration"; };
```

Comment:
I just realized `tomlFormat.generate` was used.
Is there a difference between this manual type and `tomlFormat.type`?

Reply:
I'm not 100% sure, I based the module off the existing [starship](https://github.com/nix-community/home-manager/blob/42bb553544420942415257fb2d3938698b27d72d/modules/programs/starship.nix) configuration. Open to going with whatever's the recommended option.

Reply:
I recommend using `tomlFormat.type` in that case :+1:

See this test case that shows the resulting format:
https://github.com/NixOS/nixpkgs/blob/d600f006643e074c2ef1d72e462e218b647a096c/pkgs/pkgs-lib/tests/formats.nix#L127-L154

Reply:
Done, tests passed without any changes required 🎉